src/booster/builder.rs

```diff
@@ -0,0 +1,96 @@
+use serde_json::Value;
+
+use booster::Booster;
+use dataset::DataSet;
+use Error;
+use {InputMatrix, OutputVec};
+
+// types for training set
+#[derive(Clone)]
+pub struct TrainDataAdded(DataSet); // this should not implement default, so it can safely be used for construction
+#[derive(Default, Clone)]
+pub struct TrainDataNotAdded;
```

Comment:
Thinking about the resulting API, I was wondering whether these structs could have names where the important part stands out more? Like `WithTrainData` and `NoTrainData`. The difference is the very first word, instead of (not) having a `Not` added in the middle of a fairly long type name.

Same for Params.